### PR TITLE
verify file manifest before creating batch context (if using file manifest)

### LINKED ISSUE #1117: Verify that a File Manifest includes shelve or preserve set to "yes"
If a file manifest is submitted where both "shelve" and "preserve" are set to "no" for a specific file, then the file is ultimately discarded. If a user forgets to set one of those values at all (i.e. it's left blank on the CSV) this will result in the field being set to no on deposit.

There is no use case for submitting a content file with both shelve and preserve set to no, as that would indicate that there was no need to deposit the file at all. A file_manifest with both fields set to "no" or left blank accidentally is almost certainly a mistake and the system should stop the job and report back a message to the user.

As part of the discovery report we should validate that at least one of the two fields is set to yes, which will catch both the case where both fields are set to "no" and the case where one field is set to "no" and another is left blank. 

Context: I found this issue when I made a mistake in my own accessioning job on stage and discovered that one file had been discarded.

### PR DESCRIPTION
**Pull request by peetucket**
# Why was this change made? 🤔

Fixes #1117 - verify the file manifest does not have any files with shelve and preserve both set to "no" ... this makes no sense and is not allowed.

~~The verification happens as the job is created (either discovery report or pre-assembly job) ... so you cannot even create the job if you are using a file manifest and the file itself has problems (in the same way you cannot create the job if you are using a file manifest and the file doesn't exist).~~

Also contains some refactoring around related methods, as described in PR comments in the code.

~~HOLD until integration tests are run~~
~~HOLD until verified by Andrew~~

# How was this change tested? 🤨

- [x] Localhost
- [x] Updated unit tests
- [x] Integration tests
- [x] Verified by Andrew
- [x] Works with a large file manifest

### CODE REVIEW

File: app/lib/pre_assembly/file_manifest.rb
Comment (peetucket) [inline app/lib/pre_assembly/file_manifest.rb:21]: the `exists?` method makes no sense here...because we try to load the manifest file itself when we initialize this class, and so if the file doesn't exist, the class would have already thrown an exception on initialize

an attempt to use this method was being made elsewhere in the code when checking to see if a job run was using a file manifest, but it makes more sense to just inspect the `using_file_manifest` attribue on the model, which tells you this directly

File: app/lib/pre_assembly/batch.rb
Comment (peetucket) [inline app/lib/pre_assembly/batch.rb:24]: refactor: make this database model attribute available so it can be used elsewhere in the code

File: app/lib/pre_assembly/file_manifest.rb
Comment (peetucket) [inline app/lib/pre_assembly/file_manifest.rb:28]: the new logic for verifying if the file manifest is valid...go through all files and see if of them have both preserve and shelve set to "no"

File: app/models/batch_context.rb
Comment (peetucket) [inline app/models/batch_context.rb:156]: **refactor:** the check for existence of a file manifest was moved to the new `verify_file_manifest` method added below

File: app/services/discovery_report.rb
Comment (peetucket) [inline app/services/discovery_report.rb:30]: **refactor:** removed this unused method which was using that janky `exists?` method on the file manifest object

File: app/services/object_file_validator.rb
Comment (peetucket) [inline app/services/object_file_validator.rb:112]: **refactor:** just inspect the database model object itself to decide if we are using a file manifest instead of using that janky exists? method on the file manifest object

File: spec/services/object_file_validator_spec.rb
Comment (peetucket) [inline spec/services/object_file_validator_spec.rb:42]: commented out and unneeded, just remove

**Review by ndushay — APPROVED** [review]
LGTM;  seems like integration tests pass and Andrew approved.

**Comment by peetucket** [comment]
Going to test with a very large file manifest to be sure there isn't a performance issue on form submission before merging.

**Review by jmartin-sul — APPROVED** [review]

File: app/lib/pre_assembly/file_manifest.rb
Comment (peetucket) [inline app/lib/pre_assembly/file_manifest.rb:41]: verify that there are rows in the file manifest and all required columns are present

File: spec/fixtures/multimedia/file_manifest.csv
Comment (peetucket) [inline spec/fixtures/multimedia/file_manifest.csv:17]: these existing fixtures are not actually valid anymore and need to be changed (and the ones below as well)... we cannot have "no, no, no" anymore for publish,preserve, shelve

File: spec/lib/pre_assembly/file_manifest_spec.rb
Comment (peetucket) [inline spec/lib/pre_assembly/file_manifest_spec.rb:211]: changes go with existing fixtures that were updated above (this is invalid data now and cannot be created)

File: app/models/batch_context.rb
Comment (peetucket) [inline app/models/batch_context.rb:178]: Verify that we at least have a (non-empty) file_manifest before creating the job, because that is fast...actual further validation happens once we instantiate and load the data in the `FileManifest` object (in the job, because that can be slow)

File: spec/services/object_file_validator_spec.rb
Comment (peetucket) [inline spec/services/object_file_validator_spec.rb:56]: **refactor:** remove unnecessary line of code

File: spec/lib/pre_assembly/file_manifest_spec.rb
Comment (peetucket) [inline spec/lib/pre_assembly/file_manifest_spec.rb:31]: new validation when we load the file manifest: checks for no rows, invalid rows, or missing columns

File: spec/models/batch_context_spec.rb
Comment (peetucket) [inline spec/models/batch_context_spec.rb:345]: new tests for validation when we create the batch context: file manifest exists and is not an empty file

File: spec/features/discovery_report/discovery_report_failed_spec.rb
Comment (peetucket) [inline spec/features/discovery_report/discovery_report_failed_spec.rb:40]: we get a better error now that we actually validate this properly!

**Comment by peetucket** [comment]
This PR was reworked to:

1. Move the validation of the file manifest rows (i.e. the publish/preserve combo) into the `FileManifest`class itself instead of the `BatchContext` model.  This is because if you have a very large file manifest (>10k rows), loading it into memory and validating can take a while (like 30 seconds), and if you validate this on form submission, it looks like the server is hung up while you wait for the response.   Validating it on the `FileManifest` class when loading the data means it won't happen until the job is run.  We can throw an exception if there is a problem, which will be caught by the job and put into an error state with a friendly message (this is already setup to work this way).
2. Added additional validation to not only check for file manifest existence (if selected), but also that it is not an empty file, has at least one row of data, and has the required columns.  Each of these errors would have caused problems before, but would have thrown other error messages or exceptions that were not as clear as doing this validation explicitly.

**Comment by peetucket** [comment]
Integration tests have passed and I have added/modified tests to cover the new behavior, but would good to verify with @andrewjbtw this behavior is acceptable.

**Review by jmartin-sul — APPROVED** [review]
one thought for possibly useful follow on work, but LGTM

File: app/lib/pre_assembly/file_manifest.rb
Comment (peetucket) [reply app/lib/pre_assembly/file_manifest.rb:31]: That is a good point.  Let me consider that in a new PR so we don't hold up this one.

**Pull request merged**
